src/library/scala/collection/mutable/AnyRefMap.scala

```diff
@@ -426,6 +426,15 @@ extends AbstractMap[K, V]
     this
   }
 
+  override def clear(): Unit = {
+    import java.util.Arrays.fill
+    fill(_keys, null)
+    fill(_values, null)
+    fill(_hashes, 0)
+    _size = 0
+    _vacant = 0
+  }
```

Comment:
Seems like this attempt to reused storage would mostly be foiled as the subsequent call to `put` will determine that the map is imbalanced and repack it into smaller storage?

Reply:
Looks like I'm reading the code incorrectly.

```
scala> def getFieldOf(m: Any, name: String) = { val fld = m.getClass.getDeclaredField(name); fld.setAccessible(true); fld.get(m) }
getFieldOf: (m: Any, name: String)Object

scala> def tableSizeOf(m: collection.mutable.AnyRefMap[_, _]): Int = getFieldOf(m, "scala$collection$mutable$AnyRefMap$$_keys").asInstanceOf[Array[Object]].length
tableSizeOf: (m: scala.collection.mutable.AnyRefMap[_, _])Int

scala> tableSizeOf(map)
res11: Int = 16

scala> for (i <- (1 to 100000)) { map.put(i.toString(), "") }

scala> map.size
res23: Int = 100001

scala> tableSizeOf(map)
res24: Int = 262144

scala> map.clear()

scala> tableSizeOf(map)
res26: Int = 262144

scala> map.put("", "")
res27: Option[String] = None

scala> tableSizeOf(map)
res28: Int = 262144
```

Reply:
As I understand the code, `_vacant` is the number of positions that once held an element but no longer do, so it's maintained to estimate how many entries aren't in their ideal position, not how many slots are still open. But this code hurts my eyes...

Reply:
I added a println just to get a taste of the ergonomics.

Here is `(mask,size,vacant)` bumping capacity needlessly:
```
(15,5,3) => (31,5,0)
```
I think it's because of the way vacancies are accounted when computing the new capacity, the `.2` allowance. `3` is not many, so `5` is deemed too close to load factor.

That was `transformed` updates from namers, which seems to do a lot of repacking. The promise of the code comment, that we will mutate in a building phase and then repack for fast lookups, doesn't really hold. Maybe it enjoys other efficiencies that compensate for what look like gratuitous array copies.

This was just from compiling a tiny test, nothing rigorous.

I see the pickle effect that is being fixed here. Since pickle only puts, it doesn't see vacancies. It does quickly grow, so init size could be bigger.

For other usages, I wonder if having a "minimum capacity" or a "vacancy factor" would help trade allocations for lookup speed. For example, this had more vacancies than entries, let's repack:
```
(15,1,2) => (7,1,0)
```
Maybe just use `initialBufferSize` in repack, at least, to avoid shrinkage.

Reply:
OK, I added a test that it remain like this. I don't want to fool about in `AnyRefMap` more than I have to for this PR but I'll gladly come back to break it in subtle ways. Thanks both for the analysis.